paddle/fluid/operators/conv_mkldnn_op.cc

```diff
@@ -358,132 +364,424 @@ class ConvMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
+            user_bias_memory_p, pipeline, is_test, is_INT8, scale_bias_data,
+            mask_reorder);
+        conv_p = handler.AcquireConvolution(src_memory_p, weights_memory_p,
+                                            bias_memory_p, dst_memory_p);
       } else {
-        output->ShareDataWith(*residual_param);
-        auto output_data = output->mutable_data<T>(ctx.GetPlace());
-        dst_memory_p =
-            handler.AcquireDstMemoryFromPrimitive(to_void_cast<T>(output_data));
+        conv_p = handler.AcquireConvolution(src_memory_p, weights_memory_p,
+                                            dst_memory_p);
       }
-    } else {
-      auto output_data = output->mutable_data<T>(
-          ctx.GetPlace(), paddle::memory::Allocator::kDefault,
-          handler.GetDstMemorySize());
-      dst_memory_p =
-          handler.AcquireDstMemoryFromPrimitive(to_void_cast<T>(output_data));
+
+      // push primitive to stream and wait until it's executed
+      pipeline.push_back(*conv_p);
+      stream(stream::kind::eager).submit(pipeline).wait();
+
+      if (need_s8_to_u8) {
+        output->mutable_data<uint8_t>(ctx.GetPlace());
+      }
+
+      output->set_layout(DataLayout::kMKLDNN);
+      output->set_format(GetMKLDNNFormat(*dst_memory_p));
     }
+  }
 
-    // create convolution op primitive
-    std::shared_ptr<mkldnn::convolution_forward> conv_p;
-    if (bias) {
-      const T* bias_data = bias->data<T>();
-      auto user_bias_md = platform::MKLDNNMemDesc(
-          {bias_tz}, platform::MKLDNNGetDataType<T>(), memory::format::x);
-      auto user_bias_memory_p =
-          handler.AcquireBiasMemory(user_bias_md, to_void_cast<T>(bias_data));
-
-      auto bias_memory_p =
-          handler.AcquireBiasMemoryFromPrimitive(user_bias_memory_p, pipeline);
-      conv_p = handler.AcquireConvolution(src_memory_p, weights_memory_p,
-                                          bias_memory_p, dst_memory_p);
+ private:
+  void SetScaleMap(std::unordered_map<std::string,
+                                      std::vector<float>>& scale_map,  // NOLINT
+                   const std::string& name,
+                   std::vector<float> scale_data) const {
+    auto it = scale_map.find(name);
+    if (it == scale_map.end()) {
+      scale_map[name] = scale_data;
     } else {
-      conv_p = handler.AcquireConvolution(src_memory_p, weights_memory_p,
-                                          dst_memory_p);
+      (*it).second = scale_data;
     }
+    return;
+  }
 
-    // push primitive to stream and wait until it's executed
-    pipeline.push_back(*conv_p);
-    stream(stream::kind::eager).submit(pipeline).wait();
+  std::vector<float> GetScaleMap(
+      std::unordered_map<std::string,
+                         std::vector<float>>& scale_map,  // NOLINT
+      const std::string& name) const {
+    auto it = scale_map.find(name);
+    if (it != scale_map.end()) {
+      return (*it).second;
+    }
+    return {0};
+  }
```

Comment:
You could use boost::optional to differentiate between situations when you return something meaningful, or return nothing, instead of returning empty vector.

Reply:
I was thinking about avoiding vector values with special meaning. I was thinking about something like that:

```
boost::optional<std::vector<float>> get_scale_map(
    std::unordered_map<std::string, std::vector<float>>& scale_map,
    const std::string& name) {
    auto it = scale_map.find(name);

    if (it != scale_map.end()) {
        return it->second;
    }

    return boost::none;
}
```

You could use it the following way:
```
auto a = get_scale_map(scale_map, v);
// check if a is not boost::none;
if (a) {
    // use value of a:
... = *a ...;
} 
```
What do you think about it?

Reply:
Similar to above comments.